File: csharp/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0)
        {
            throw new ArgumentOutOfRangeException();
        }
        var sum = Factors(number).Sum();
        if (sum == number)
        {
            return Classification.Perfect;
        }
        if (sum > number)
        {
            return Classification.Abundant;
        }
        else
        {
            return Classification.Deficient;
        }
    }

    public static List<int> Factors(int m) => Enumerable.Range(1, m / 2).Where(n => m % n == 0).ToList();
}

File: elixir/perfect_numbers.ex
defmodule PerfectNumbers do
  @type number_kind :: :perfect | :abundant | :deficient
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, number_kind} | {:error, String.t()}
  def classify(n) when n <= 0 do
    {:error, "Classification is only possible for natural numbers."}
  end

  def classify(1), do: {:ok, :deficient}

  def classify(number) do
    number
    |> aliquot_sum()
    |> check_kind(number)
  end

  def aliquot_sum(n) do
    1..div(n, 2)
    |> Enum.filter(&(rem(n, &1) == 0))
    |> Enum.sum()
  end

  def check_kind(s, n) when s > n, do: {:ok, :abundant}
  def check_kind(s, n) when s == n, do: {:ok, :perfect}
  def check_kind(_s, _n), do: {:ok, :deficient}
end
